### PR TITLE
Remove hard dependency on ext-mbstring and ext-fileinfo

### PR DESCRIPTION
**Pull request by nicolas-grekas**
This PR is aimed at increasing the portability of the package.

### CODE REVIEW

File: src/Uri.php [hunk 1]
```diff
@@ -671,10 +679,15 @@ public static function createFromComponents(array $components = []): self
      *
      * @param resource|null $context
      *
-     * @throws SyntaxError If the file does not exist or is not readable
+     * @throws IdnSupportMissing If ext/fileinfo is not installed
```
Comment (andrew-demb) [inline src/Uri.php]: `IdnSupportMissing -> FileinfoSupportMissing`

Comment (nicolas-grekas) [reply src/Uri.php]: fixed thanks.

File: src/Uri.php [hunk 2]
```diff
@@ -671,10 +679,15 @@ public static function createFromComponents(array $components = []): self
      *
      * @param resource|null $context
      *
-     * @throws SyntaxError If the file does not exist or is not readable
+     * @throws FileinfoSupportMissing If ext/fileinfo is not installed
+     * @throws SyntaxError            If the file does not exist or is not readable
      */
     public static function createFromDataPath(string $path, $context = null): self
     {
+        if (!class_exists(\finfo::class)) {
```
Comment (nyamsprod) [inline src/Uri.php]: Question why not create a static variable so to make the check only once something along the line of what I did for the IDN support check 

```php
static $finfo_support = null;
$finfo_support = $finfo_support ?? class_exists(\finfo::class);

if (!$finfo_support) {
  throw ...
}
```

Comment (nicolas-grekas) [reply src/Uri.php]: works for me, updated.

**Comment by nyamsprod** [comment]
thanks for the dependency removal. I believe this needs a minor release 🤔 since we are introducing a new exception and updating the package dependencies. I'll try to release the new version next week.

**Comment by nyamsprod** [comment]
After reviewing the newly added exception I'll migrate it to the uri-interfaces package so that it can be used in the uri-components  package too. In the end, all three packages will be minor bumped which might take little more time but it is worth it IMHO

**Comment by nyamsprod** [comment]
the 6.2.0 release is out ... faster than I thought 😄 . While porting the exception to the uri-component I notice that I had already remove the mb_detect_encoding part and replaced it with a regular expression https://github.com/thephpleague/uri-components/blob/master/src/Components/DataPath.php#L135 . Since I consider it to be just an implementation detail I did not change the code but out of curiosity I'm wondering which code is faster. I should benchmark both implementation when I'll have time

the regular expression used is the following `REGEXP_NON_ASCII_PATTERN = '/[^\x20-\x7f]/';`

**Comment by nicolas-grekas** [comment]
I would bet on strspn()
But you might want to remove some characters from the ASCII const (and rename it then)

**Comment by nyamsprod** [comment]
then I'll port the full code to uri-component then ... it's just a fix ... I'll probably do it during the week  😉 . thanks

**Pull request merged**